docs/reference/api/platform/FileHandle.md

```diff
@@ -0,0 +1,274 @@
+#include "mbed.h"
+
+static DigitalOut led1(LED1);
+static DigitalOut led2(LED2);
+
+static UARTSerial device(MBED_CONF_APP_DEVICE_TX, MBED_CONF_APP_DEVICE_RX);
+
+static void callback_ex()
+{
+    // always read until data is exhausted - we may not get another
+    // sigio otherwise
+    while (1) {
+        char c;
+        if (device.read(&c, 1) != 1) {
+            break;
+        }
+        putchar(c);
+        led2 = !led2;
+    }
+}
+
+int main()
+{
+    // UARTSerial-specific method - all others are from FileHandle base class
+    device.set_baud(19200);
+    
+    // Ensure that device.read() returns -EAGAIN when out of data
+    device.set_blocking(false);
+    
+    // sigio callback is deferred to event queue, as we cannot in general
+    // perform read() calls directly from the sigio() callback.
+    device.sigio(mbed_event_queue()->event(callback_ex));
+
+    while (1) {
+        led1 = !led1;
+        wait(0.5);
+    }
+}
+
+```
+
+### Poll example
+```
+// Transfer bidirectional data between two ports, acting as a virtual link.
+// poll() is used to monitor both ports for input.
+#include "mbed.h"
+
+// Pins for each port are specified via mbed_app.json. We assume no flow control.
+// (If there was flow control, being blocked by it could break the implementation,
+// as we would stop reading input in the opposite direction).
+UARTSerial device1(MBED_CONF_APP_UART1_TX, MBED_CONF_APP_UART1_RX);
+UARTSerial device2(MBED_CONF_APP_UART2_TX, MBED_CONF_APP_UART2_RX);
+
+// Precondition: "in" is readable
+static void copy_some(FileHandle *out, FileHandle *in)
+{
+    // To ensure performance, we try to read multiple bytes at once, although
+    // we don't expect to read many in practice.
+    char buffer[32];
+
+    // Despite the FileHandle being in its default blocking mode,
+    // read() must return immediately with between 1-32 bytes, as
+    // we've already checked that `in` is ready with poll()
+    ssize_t read = in->read(buffer, sizeof buffer);
+    if (read <= 0) {
+        error("Input error");
+    }
+
+    // Write everything out. Assuming output port is same speed as input,
+    // and no flow control, this may block briefly, but not significantly.
+    ssize_t written = out->write(buffer, read);
+    if (written < read) {
+        error("Output error");
+    }
+}
+
+int main()
+{
+    char buffer[32];
+    pollfh fds[2];
+
+    fds[0].fh = &device1;
+    fds[0].events = POLLIN;
+    fds[1].fh = &device2;
+    fds[1].events = POLLIN;
+
+    while (1) {
+        // Block indefinitely until either of the 2 ports is readable (or has an error)
+        poll(fds, 2, -1);
+
+        // Transfer some data in either or both directions
+        if (fds[0].revents) {
+            copy_some(fds[1].fh, fds[0].fh);
+        }
+        if (fds[1].revents) {
+            copy_some(fds[0].fh, fds[1].fh);
+        }
+    }
+}
+```
+
+### Related content
+
+- [File](file.html).
+- [FileSystem](filesystem.html).
```

Comment:
Maybe add UARTSerial here?

Eventually it would be nice to make a "implementation" list section similar to the blockdevices. Though we're probably not there yet with a single non-filesystem implementation :laughing: